numpy/random/mtrand/mtrand.pyx

```diff
@@ -1159,22 +1159,30 @@ cdef class RandomState:
         """
         return disc0_array(self.internal_state, rk_long, size, self.lock)
 
-    def randint(self, low, high=None, size=None, dtype='l'):
+    def randint(self, low=None, high=None, size=None, dtype='l'):
         """
-        randint(low, high=None, size=None, dtype='l')
+        randint(low=None, high=None, size=None, dtype='l')
 
         Return random integers from `low` (inclusive) to `high` (exclusive).
 
         Return random integers from the "discrete uniform" distribution of
-        the specified dtype in the "half-open" interval [`low`, `high`). If
-        `high` is None (the default), then results are from [0, `low`).
+        the specified dtype in the "half-open" interval [`low`, `high`).
+
+        If ``low`` and ``high`` are None (default), the results are from
+        [`lowbnd`, `highbnd`), where `lowbnd` and `highbnd` are equal to
+        `np.iinfo(dtype).min` and `np.iinfo(dtype).max`.
+
+        If only ``low`` is None, the results are from [`lowbnd`, `high`),
+        where `lowbnd` is defined as previously. If only `high` is None,
+        then if `low` > 0, results are from [0, `low`). This behaviour is
+        intended to maintain **backwards compatibility**. Otherwise, results
+        are from [`low`, `highbnd`), where `highbnd` is defined as previously.
```

Comment:
The latter explanation is confusing IMO: it seems very out of place when you compare it to the case when `low=None` and `high != None`.  Backwards compatibility is the sole reason for me leaving the behavior as it is.  Otherwise, I would have changed it.

Probably many will still use `randint` with position arguments, but nevertheless people should be encouraged to adapt to the new API since it technically isn't breaking.  I'm not sure I follow your reasoning as to why function calls in the code-base should continue to follow positional arguments when using `=` is a lot cleaner and clearer just because of ingrained user habits.